### PR TITLE
ceph-disk/ceph_disk/main.py: Replace ST_ISBLK() test by is_diskdevice()

### PR DESCRIPTION
**Pull request by wjwithagen**
- FreeBSD does not have blockdevices any more (since 2002)
   So disk are just Character special devices, so test on ISCHR

Signed-off-by: Willem Jan Withagen <wjw@digiware.nl>

### CODE REVIEW

File: src/ceph-disk/ceph_disk/main.py [hunk 1]
```diff
@@ -2732,10 +2738,10 @@ def set_type(self):
 
         if stat.S_ISDIR(dmode):
             self.type = self.FILE
-        elif stat.S_ISBLK(dmode):
+        elif is_diskdevice(self.args.data):
             self.type = self.DEVICE
         else:
-            raise Error('not a dir or block device', self.args.data)
+            raise Error('not a dir or block/disk device', self.args.data)
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: "not a dir or device" ?

File: src/ceph-disk/ceph_disk/main.py [hunk 2]
```diff
@@ -4059,9 +4064,8 @@ def get_space_osd_uuid(name, path):
     if not os.path.exists(path):
         raise Error('%s does not exist' % path)
 
-    mode = os.stat(path).st_mode
-    if not stat.S_ISBLK(mode):
-        raise Error('%s is not a block device' % path)
+    if not is_diskdevice(path):
+        raise Error('%s is not a block/disk device' % path)
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: "not a device"

File: src/ceph-disk/ceph_disk/main.py [hunk 3]
```diff
@@ -4664,8 +4668,8 @@ def set_suppress(path):
     disk = os.path.realpath(path)
     if not os.path.exists(disk):
         raise Error('does not exist', path)
-    if not stat.S_ISBLK(os.lstat(path).st_mode):
-        raise Error('not a block device', path)
+    if not is_diskdevice(path):
+        raise Error('not a block/disk device', path)
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: "not a device"

File: src/ceph-disk/ceph_disk/main.py [hunk 4]
```diff
@@ -4677,8 +4681,8 @@ def unset_suppress(path):
     disk = os.path.realpath(path)
     if not os.path.exists(disk):
         raise Error('does not exist', path)
-    if not stat.S_ISBLK(os.lstat(path).st_mode):
-        raise Error('not a block device', path)
+    if not is_diskdevice(disk):
+        raise Error('not a block/disk device', path)
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: "not a device"

File: src/ceph-disk/ceph_disk/main.py [hunk 5]
```diff
@@ -4677,8 +4681,8 @@ def unset_suppress(path):
     disk = os.path.realpath(path)
     if not os.path.exists(disk):
         raise Error('does not exist', path)
-    if not stat.S_ISBLK(os.lstat(path).st_mode):
-        raise Error('not a block device', path)
+    if not is_diskdevice(disk):
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: s/disk/path/

File: src/ceph-disk/ceph_disk/main.py [hunk 6]
```diff
@@ -665,6 +665,19 @@ def get_dev_size(dev, size='megabytes'):
         os.close(fd)
 
 
+def is_diskdevice
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: is_device ? I don't think we deal with anything else but devices that contain bytes

File: src/ceph-disk/ceph_disk/main.py [hunk 7]
```diff
@@ -665,6 +665,19 @@ def get_dev_size(dev, size='megabytes'):
...
+    dev = os.path.realpath(dev)
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: this must reside outside of the function: it matters to the caller and makes a difference when it is gone

File: src/ceph-disk/ceph_disk/main.py (hunk 6 above)
Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
I'm open to suggestions for another function name. I though `is_blockdevice` but then that is exactly what it is not. It is the bare metal thingy ultimately, the diskdevice.

File: src/ceph-disk/ceph_disk/main.py (hunk 7 above)
Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
Does it really matter for this function? I mean, it is a real device, or a link to the device.
Do you have cases where it is a link to a device, but you still would like the answer to this function to be false? Could be, given my limited knowledge of all the variations on linux devices

File: src/ceph-disk/ceph_disk/main.py (hunk 1 above)
Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
You mean: just kill the  'block/disk' part. Can do that...

File: src/ceph-disk/ceph_disk/main.py (hunk 6 above)
Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
I also see things happening on files. That is certainly not ISBLK.
Also there are in Filestore directories on which to build the Ceph stuff.
Again not ISBLK I think?

File: src/ceph-disk/ceph_disk/main.py (hunk 1 above)
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: yes

File: src/ceph-disk/ceph_disk/main.py (hunk 6 above)
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: Right. We either have files, directories or devices in the sense that they are in /dev or are special files.

File: src/ceph-disk/ceph_disk/main.py (hunk 7 above)
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: It does matter. see below.

File: src/ceph-disk/ceph_disk/main.py [hunk 8]
```diff
@@ -1536,9 +1546,7 @@ def zap(dev):
     """
     Destroy the partition table and content of a given disk.
     """
-    dev = os.path.realpath(dev)
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: You must keep dev here. It is used later in the function for display  and processing. https://docs.python.org/2/library/os.path.html#os.path.realpath removes symbolic links. This is the part I'm worried about because sometime we want the symbolic link, not the device it points to.

os.path.realpath is used carefully and it is complicated to figure out exactly why and where. This is why it needs to stay exactly where it is and not be used more than it should.

Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
Oke, I'll go over the PR with a fine comb, because I think I also killed af few `stat()` versus `lstat()`, and that might then also matter. I'll put back the leader part of the function in the code.
I thought that it was also a nice cleanup as extra bonus, but obviously not.

**Comment by wjwithagen** [comment]
@dachary 
Ping?

File: src/ceph-disk/ceph_disk/main.py [hunk 9]
```diff
@@ -777,8 +803,8 @@ def get_partition_base(dev):
     Get the base device for a partition
     """
     dev = os.path.realpath(dev)
-    if not stat.S_ISBLK(os.lstat(dev).st_mode):
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: the lstat is useless here: os.path.realpath will return the actual file, not the symlink, therefore dev will never be a symlink

File: src/ceph-disk/ceph_disk/main.py [hunk 10]
```diff
@@ -4649,7 +4672,7 @@ def is_suppressed(path):
     disk = os.path.realpath(path)
     try:
         if (not disk.startswith('/dev/') or
-                not stat.S_ISBLK(os.lstat(disk).st_mode)):
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: lstat useless here as well, stat will do the same because of the realpath above

File: src/ceph-disk/ceph_disk/main.py [hunk 11]
```diff
@@ -4677,8 +4700,8 @@ def unset_suppress(path):
     disk = os.path.realpath(path)
     if not os.path.exists(disk):
         raise Error('does not exist', path)
-    if not stat.S_ISBLK(os.lstat(path).st_mode):
-        raise Error('not a block device', path)
+    if not ldev_is_diskdevice(path):
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: I think we can change path to disk here, for consistency. I don't see any useful use case where it would matter to use path instead of disk. This is mostly likely what was intended originally.

Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: and use dev_is_diskdevice

Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
Wiil go over your careful review, and fix. Thanx ...

**Comment by wjwithagen** [comment]
@dachary 
Looks like the error is known: cephtool-test-mon.sh

**Comment by wjwithagen** [comment]
@dachary 
Ping?

**Comment by dachary** [comment]
jenkins test this please (master is fixed)

**Comment by wjwithagen** [comment]
@dachary 
Jenkins seems happy. Do you want to run it thru QA?

File: src/ceph-disk/ceph_disk/main.py (hunk 11 above)
Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
Hi Loic, could you take a look again, and merge if oke?

**Comment by wjwithagen** [comment]
@yuriw 
Sorry about the merge conflict. PRs went in, in the "wrong" order. 
Should be fixed now.

**Comment by liewegas** [comment]
retest this please

File: src/ceph-disk/tests/ceph-disk.sh [hunk 12]
```diff
@@ -205,7 +205,7 @@ function test_zap() {
     local osd_data=$dir/dir
     $mkdir -p $osd_data
 
-    ${CEPH_DISK} $CEPH_DISK_ARGS zap $osd_data 2>&1 | grep -q 'not full block device' || return 1
+    ${CEPH_DISK} $CEPH_DISK_ARGS zap $osd_data 2>&1 | grep -q 'not full device' || return 1
```
Comment (tchaikov) [inline src/ceph-disk/tests/ceph-disk.sh]: @wjwithagen why did you change the grepped error message?

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-disk/tests/ceph-disk.sh:208: test_zap:  grep -q 'not full device'
/home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-disk/tests/ceph-disk.sh:208: test_zap:  /home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-disk/.tox/py27/bin/coverage run --append --source=ceph_disk -- /home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-disk/.tox/py27/bin/ceph-disk --verbose --prepend-to-path= --statedir=td/test-ceph-disk --sysconfdir=td/test-ceph-disk zap td/test-ceph-disk/dir
/home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-disk/tests/ceph-disk.sh:208: test_zap:  return 1
```

and it looks like "make check" failed due to this change, see https://jenkins.ceph.com/job/ceph-pull-requests/28056/consoleFull#-23068421240526d21-3511-427d-909c-dd086c0d1034.

Comment (wjwithagen) [inline src/ceph-disk/tests/ceph-disk.sh]: @tchaikov 
I changed the grepped messages because the test no longer is only for a block device. 
FreeBSD does not have block devices, so not to "confuse" user, and get useless questions, I thought it was wise to make it more generic. But perhaps I missed some tests.
And things could get worse when goint into theutology...
So perhaps it is easier to just change this back.

**Comment by wjwithagen** [comment]
@tchaikov 
Reverted the error messages, lets see if this passes Jenkins and QA

**Comment by tchaikov** [comment]
retest this please.

File: src/ceph-disk/tests/ceph-disk.sh (hunk 12 above)
Comment (tchaikov) [inline src/ceph-disk/tests/ceph-disk.sh]: if you believe that "block device" is not accurate or wrong on FreeBSD, probably you can update where the exception is raised. or better off conditionalize this on FreeBSD/Linux, so Linux user can have more specific warning message.

File: src/ceph-disk/ceph_disk/main.py [hunk 13]
```diff
@@ -819,7 +845,7 @@ def is_partition(dev):
 
     dev = os.path.realpath(dev)
     st = os.lstat(dev)
-    if not stat.S_ISBLK(st.st_mode):
+    if not stmode_is_diskdevice(st.mode):
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: s/st.mode/st.st_mode/
```
2017-07-13T16:47:20.982 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk::sh: ceph-disk list --format json vda vdb vdc vdd
2017-07-13T16:47:21.068 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:Traceback (most recent call last):
2017-07-13T16:47:21.069 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/sbin/ceph-disk", line 9, in <module>
2017-07-13T16:47:21.069 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:load_entry_point('ceph-disk==1.0.0', 'console_scripts', 'ceph-disk')()
2017-07-13T16:47:21.069 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 5725, in run
2017-07-13T16:47:21.069 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:main(sys.argv[1:])
2017-07-13T16:47:21.069 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 5678, in main
2017-07-13T16:47:21.070 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:main_catch(args.func, args)
2017-07-13T16:47:21.070 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 5703, in main_catch
2017-07-13T16:47:21.070 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:func(args)
2017-07-13T16:47:21.070 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 4638, in main_list
2017-07-13T16:47:21.071 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:main_list_protected(args)
2017-07-13T16:47:21.071 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 4642, in main_list_protected
2017-07-13T16:47:21.071 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:devices = list_devices()
2017-07-13T16:47:21.071 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 4602, in list_devices
2017-07-13T16:47:21.072 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:device = list_dev(get_dev_path(base), uuid_map, space_map)
2017-07-13T16:47:21.073 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 4473, in list_dev
2017-07-13T16:47:21.073 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:info['is_partition'] = is_partition(dev)
2017-07-13T16:47:21.073 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 869, in is_partition
2017-07-13T16:47:21.073 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:if not stmode_is_diskdevice(st.mode):
2017-07-13T16:47:21.074 INFO:tasks.workunit.client.0.vpm073.stderr:DEBUG:CephDisk:AttributeError: 'posix.stat_result' object has no attribute 'mode'
2017-07-13T16:47:21.094 INFO:tasks.workunit.client.0.vpm073.stdout:../../../clone.client.0/qa/workunits/ceph-disk/ceph-disk-test.py:345: TestCephDisk.test_activate_dmcrypt_luks_wit
h_lockbox FAILED
```

File: src/ceph-disk/ceph_disk/main.py [hunk 14]
```diff
@@ -777,7 +803,7 @@ def get_partition_base(dev):
     Get the base device for a partition
     """
     dev = os.path.realpath(dev)
-    if not stat.S_ISBLK(os.lstat(dev).st_mode):
+    if ldev_is_diskdevice(dev):
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: should be 
```py
if not ldev_is_diskdevice(dev):
```

```
2017-07-14T04:09:31.086 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:Traceback (most recent call last):
2017-07-14T04:09:31.088 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:File "/usr/sbin/ceph-disk", line 9, in <module>
2017-07-14T04:09:31.091 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:load_entry_point('ceph-disk==1.0.0', 'console_scripts', 'ceph-disk')()
2017-07-14T04:09:31.113 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 5725, in run
2017-07-14T04:09:31.116 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:main(sys.argv[1:])
2017-07-14T04:09:31.125 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 5676, in main
2017-07-14T04:09:31.127 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:args.func(args)
2017-07-14T04:09:31.130 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 4034, in main_destroy
2017-07-14T04:09:31.198 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:main_destroy_locked(args)
2017-07-14T04:09:31.236 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 4061, in main_destroy_locked
2017-07-14T04:09:31.239 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:base_dev = get_partition_base(dev_path)
2017-07-14T04:09:31.242 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:File "/usr/lib/python2.7/dist-packages/ceph_disk/main.py", line 828, in get_partition_base
2017-07-14T04:09:31.244 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:raise Error('not a block device', dev)
2017-07-14T04:09:31.250 INFO:tasks.workunit.client.0.vpm021.stderr:DEBUG:CephDisk:ceph_disk.main.Error: Error: not a block device: /dev/vdb1
```

/a/kchai-2017-07-14_03:23:35-ceph-disk-wip-15432-kefu-distro-basic-vps/1398707/teuthology.log

File: src/ceph-disk/ceph_disk/main.py [hunk 15]
```diff
@@ -665,6 +665,32 @@ def get_dev_size(dev, size='megabytes'):
         os.close(fd)
 
 
+def stmode_is_diskdevice(dmode):
+    if stat.S_ISBLK(dmode):
+        return True
+    else:
+        # FreeBSD does not have block devices
+        # All disks are character devices
+        if FREEBSD and stat.S_ISCHR(dmode):
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: nit, could just put
```py
    return FREEBSD and stat.S_ISCHR(dmode)
```
here

**Comment by wjwithagen** [comment]
@tchaikov 
Thanx for running and explaining the theutology bugs. I always have a hard time finding the actual mistake in the sea of lines.
Submitted  fixes, running thru Jenkins.

**Comment by tchaikov** [comment]
ceph-disk suite fails on master: http://pulpito.ceph.com/kchai-2017-07-14_16:33:13-ceph-disk-master-distro-basic-vps/

we need to fixed it before running this PR through the test suite.

**Comment by wjwithagen** [comment]
@tchaikov 
Looks like something that has to do with a missing `global p`??
Doesn't llok like something I intentionally modified.

**Comment by tchaikov** [comment]
@wjwithagen i think we need to whitelist some health warnings. why do you think we need to have a `global p` and where should we have it?

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-07-15_03:18:28-ceph-disk-wip-15432-kefu-distro-basic-vps/

**Comment by tchaikov** [comment]
my bad: the test above did not contain this PR .
